In [22]:
import re
from test_utilities import run_tests_params

<link href="style.css" rel="stylesheet"></link>
<main>
<article class="day-desc"><h2>--- Day 1: Not Quite Lisp ---</h2><p>Santa was hoping for a white Christmas, but his weather machine's "snow" function is powered by stars, and he's fresh out!  To save Christmas, he needs you to collect <em class="star">fifty stars</em> by December 25th.</p>
<p>Collect stars by helping Santa solve puzzles.  Two puzzles will be made available on each day in the Advent calendar; the second puzzle is unlocked when you complete the first.  Each puzzle grants <em class="star">one star</em>. <span title="Also, some puzzles contain Easter eggs like this one. Yes, I know it's not traditional to do Advent calendars for Easter.">Good luck!</span></p>
<p>Here's an easy puzzle to warm you up.</p>
<p>Santa is trying to deliver presents in a large apartment building, but he can't find the right floor - the directions he got are a little confusing. He starts on the ground floor (floor <code>0</code>) and then follows the instructions one character at a time.</p>
<p>An opening parenthesis, <code>(</code>, means he should go up one floor, and a closing parenthesis, <code>)</code>, means he should go down one floor.</p>
<p>The apartment building is very tall, and the basement is very deep; he will never find the top or bottom floors.</p>
<p>For example:</p>
<ul>
<li><code>(())</code> and <code>()()</code> both result in floor <code>0</code>.</li>
<li><code>(((</code> and <code>(()(()(</code> both result in floor <code>3</code>.</li>
<li><code>))(((((</code> also results in floor <code>3</code>.</li>
<li><code>())</code> and <code>))(</code> both result in floor <code>-1</code> (the first basement level).</li>
<li><code>)))</code> and <code>)())())</code> both result in floor <code>-3</code>.</li>
</ul>
<p>To <em>what floor</em> do the instructions take Santa?</p>
</article>
</main>


In [25]:
tests = [
    {
        "name": "Example 1",
        "num": "(())",
        "expected": 0,
    },
    {
        "name": "Example 2",
        "num": "()()",
        "expected": 0,
    },
    {
        "name": "Example 3",
        "num": "(((",
        "expected": 3,
    },
    {
        "name": "Example 4",
        "num": "(()(()(",
        "expected": 3,
    },
    {
        "name": "Example 5",
        "num": "))(((((",
        "expected": 3,
    },
    {
        "name": "Example 6",
        "num": "())",
        "expected": -1,
    },
    {
        "name": "Example 7",
        "num": "))(",
        "expected": -1,
    },
    {
        "name": "Example 8",
        "num": ")))",
        "expected": -3,
    },
    {
        "name": "Example 8",
        "num": ")())())",
        "expected": -3,
    },
]


def floor(num: str) -> int:
    return num.count("(") - num.count(")")


run_tests_params(floor, tests)


Test Example 1 passed, for floor.
Test Example 2 passed, for floor.
Test Example 3 passed, for floor.
Test Example 4 passed, for floor.
Test Example 5 passed, for floor.
Test Example 6 passed, for floor.
Test Example 7 passed, for floor.
Test Example 8 passed, for floor.
Test Example 8 passed, for floor.
Success


In [26]:
with open("../input/day1.txt") as f:
    print(floor(f.read()))

280


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>280</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>Now, given the same instructions, find the <em>position</em> of the first character that causes him to enter the basement (floor <code>-1</code>).  The first character in the instructions has position <code>1</code>, the second character has position <code>2</code>, and so on.</p>
<p>For example:</p>
<ul>
<li><code>)</code> causes him to enter the basement at character position <code>1</code>.</li>
<li><code>()())</code> causes him to enter the basement at character position <code>5</code>.</li>
</ul>
<p>What is the <em>position</em> of the character that causes Santa to first enter the basement?</p>
</article>

</main>


In [28]:
tests = [
    {
        "name": "Example 1",
        "num": ")",
        "expected": 1,
    },
    {
        "name": "Example 2",
        "num": "()())",
        "expected": 5,
    },
]


# class Solution:
#     @staticmethod


def floor(num: str) -> int:
    count = 0
    for i, c in enumerate(num, start=1):
        if c == "(":
            count += 1
        else:
            count -= 1
            if count == -1:
                return i

    return -1


run_tests_params(floor, tests)


Test Example 1 passed, for floor.
Test Example 2 passed, for floor.
Success


In [29]:
with open("../input/day1.txt") as f:
    print(floor(f.read()))

1797


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>1797</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>
<p>At this point, you should <a href="/2015">return to your Advent calendar</a> and try another puzzle.</p>
<p>If you still want to see it, you can <a href="1/input" target="_blank">get your puzzle input</a>.</p>
<p>You can also <span class="share">[Share<span class="share-content">on
  <a href="https://twitter.com/intent/tweet?text=I%27ve+completed+%22Not+Quite+Lisp%22+%2D+Day+1+%2D+Advent+of+Code+2015&amp;url=https%3A%2F%2Fadventofcode%2Ecom%2F2015%2Fday%2F1&amp;related=ericwastl&amp;hashtags=AdventOfCode" target="_blank">Twitter</a>
  <a href="javascript:void(0);" onclick="var ms; try{ms=localStorage.getItem('mastodon.server')}finally{} if(typeof ms!=='string')ms=''; ms=prompt('Mastodon Server?',ms); if(typeof ms==='string' &amp;&amp; ms.length){this.href='https://'+ms+'/share?text=I%27ve+completed+%22Not+Quite+Lisp%22+%2D+Day+1+%2D+Advent+of+Code+2015+%23AdventOfCode+https%3A%2F%2Fadventofcode%2Ecom%2F2015%2Fday%2F1';try{localStorage.setItem('mastodon.server',ms);}finally{}}else{return false;}" target="_blank">Mastodon</a></span>]</span> this puzzle.</p>
</main>
